search/views.py

```diff
@@ -318,12 +123,154 @@ def search(request):
+
+    # check if computation is finished. If not, send computation state.
+    if result['finished']:
+        if result['result'] is not None:
+            results = result['result']
+            num_clusters = len(results)
+        else:
+             return JsonResponse({'status': 'failed'}, safe=False)
+    elif result['error']:
+        return JsonResponse({'status': 'failed'}, safe=False)
+    else:
+        return JsonResponse({'status': 'pending'}, safe=False)
+
+    # check if facet filters are present in the search query
+    # if yes, filter sounds from clusters
+    query_params, _, _ = search_prepare_parameters(request)
+    if query_params['filter_query']:
+        sound_ids_filtered = get_sound_ids_from_solr_query(query_params)
+        results = [[sound_id for sound_id in cluster if int(sound_id) in sound_ids_filtered] 
+                   for cluster in results]
+
+    num_sounds_per_cluster = [len(cluster) for cluster in results]
+    classes = {sound_id: cluster_id for cluster_id, cluster in enumerate(results) for sound_id in cluster}
+
+    # label clusters using most occuring tags
+    sound_instances = sounds.models.Sound.objects.bulk_query_id(map(int, classes.keys()))
+    sound_tags = {sound.id: sound.tag_array for sound in sound_instances}
+    cluster_tags = defaultdict(list)
+
+    # extract tags for each clusters and do not use query terms for labeling clusters
+    query_terms = {t.lower() for t in request.GET.get('q', '').split(' ')}
+    for sound_id, tags in sound_tags.iteritems():
+        cluster_tags[classes[str(sound_id)]] += [t.lower() for t in tags if t.lower() not in query_terms]
+
+    # count 3 most occuring tags
+    # we iterate with range(len(results)) to ensure that we get the right order when iterating through the dict 
+    cluster_most_occuring_tags = [' '.join(zip(*Counter(cluster_tags[cluster_id]).most_common(3))[0]) 
+                                  if len(cluster_tags[cluster_id])>0 else ''
+                                  for cluster_id in range(len(results))]
+
+    # extract sound examples for each cluster
+    sound_ids_examples_per_cluster = [map(int, cluster_sound_ids[:7]) for cluster_sound_ids in results]
+    sound_ids_examples = [item for sublist in sound_ids_examples_per_cluster for item in sublist]
+    sound_urls = {
+        sound.id: sound.locations()['preview']['LQ']['ogg']['url'] 
+            for sound in sound_instances 
+            if sound.id in sound_ids_examples
+    }
+    sound_url_examples_per_cluster = [
+        [(sound_id, sound_urls[sound_id]) for sound_id in cluster_sound_ids] 
+            for cluster_sound_ids in sound_ids_examples_per_cluster
+    ]
+
+    return render(request, 'search/clustering_facet.html', {
+            'results': classes,
+            'url_query_params_string': url_query_params_string,
+            'cluster_id_num_results': zip(range(num_clusters), num_sounds_per_cluster, 
```

Comment:
It might be nicer to be a bit more explicit here. Could you make 'cluster_id_num_results' a dictionary, with each of these 4 items as values?